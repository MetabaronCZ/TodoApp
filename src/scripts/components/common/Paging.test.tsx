import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { Paging } from 'components/common/Paging';
import { withMockedProviders } from 'test/component';

describe('components/common/Paging', () => {
  it('should render correctly', () => {
    const onChange = vi.fn();

    const tree = render(
      withMockedProviders(
        <Paging page={11} perPage={10} count={500} onChange={onChange} />,
      ),
    );
    expect(tree.container).toMatchSnapshot();
    expect(tree.getByText('12')).toBeInTheDocument();
    expect(tree.getAllByRole('button').length).toEqual(2);
  });

  it('should render one page paging', () => {
    const onChange = vi.fn();

    const tree = render(
      withMockedProviders(
        <Paging page={0} perPage={10} count={5} onChange={onChange} />,
      ),
    );
    expect(tree.container).toMatchSnapshot();
    expect(tree.getByText('1')).toBeInTheDocument();

    const buttons = tree.getAllByRole('button');
    expect(buttons[0]).toHaveAttribute('disabled');
    expect(buttons[1]).toHaveAttribute('disabled');
  });

  it('should render first page', () => {
    const onChange = vi.fn();

    const tree = render(
      withMockedProviders(
        <Paging page={0} perPage={10} count={25} onChange={onChange} />,
      ),
    );
    expect(tree.container).toMatchSnapshot();
    expect(onChange).not.toBeCalled();

    const buttons = tree.getAllByRole('button');
    expect(buttons[0]).toHaveAttribute('disabled');
    expect(buttons[1]).not.toHaveAttribute('disabled');

    fireEvent.click(buttons[1]);
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).lastCalledWith(1);
  });

  it('should render last page', () => {
    const onChange = vi.fn();

    const tree = render(
      withMockedProviders(
        <Paging page={2} perPage={10} count={25} onChange={onChange} />,
      ),
    );
    expect(tree.container).toMatchSnapshot();
    expect(onChange).not.toBeCalled();

    const buttons = tree.getAllByRole('button');
    expect(buttons[0]).not.toHaveAttribute('disabled');
    expect(buttons[1]).toHaveAttribute('disabled');

    fireEvent.click(buttons[0]);
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).lastCalledWith(1);
  });

  it('should render inner page', () => {
    const onChange = vi.fn();

    const tree = render(
      withMockedProviders(
        <Paging page={1} perPage={10} count={25} onChange={onChange} />,
      ),
    );
    expect(tree.container).toMatchSnapshot();
    expect(onChange).not.toBeCalled();

    const buttons = tree.getAllByRole('button');
    expect(buttons[0]).not.toHaveAttribute('disabled');
    expect(buttons[1]).not.toHaveAttribute('disabled');

    fireEvent.click(buttons[0]);
    fireEvent.click(buttons[1]);
    expect(onChange).toBeCalledTimes(2);
    expect(onChange).nthCalledWith(1, 0);
    expect(onChange).nthCalledWith(2, 2);
  });
});
