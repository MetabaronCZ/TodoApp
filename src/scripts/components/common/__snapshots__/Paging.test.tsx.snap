// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/common/Paging > should render correctly 1`] = `
.c3 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c2:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 10px;
  border-top: 1px solid #c0c0c0;
}

.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
      title="Předchozí"
      type="button"
    >
      <i
        class="c3"
      >
        ❮
      </i>
    </button>
    <div
      class="c4"
    >
      12
    </div>
    <button
      class="c1 c2"
      title="Další"
      type="button"
    >
      <i
        class="c3"
      >
        ❯
      </i>
    </button>
  </div>
</div>
`;

exports[`components/common/Paging > should render first page 1`] = `
.c3 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c2:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 10px;
  border-top: 1px solid #c0c0c0;
}

.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
      disabled=""
      title="Předchozí"
      type="button"
    >
      <i
        class="c3"
        style="color: rgb(192, 192, 192);"
      >
        ❮
      </i>
    </button>
    <div
      class="c4"
    >
      1
    </div>
    <button
      class="c1 c2"
      title="Další"
      type="button"
    >
      <i
        class="c3"
      >
        ❯
      </i>
    </button>
  </div>
</div>
`;

exports[`components/common/Paging > should render inner page 1`] = `
.c3 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c2:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 10px;
  border-top: 1px solid #c0c0c0;
}

.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
      title="Předchozí"
      type="button"
    >
      <i
        class="c3"
      >
        ❮
      </i>
    </button>
    <div
      class="c4"
    >
      2
    </div>
    <button
      class="c1 c2"
      title="Další"
      type="button"
    >
      <i
        class="c3"
      >
        ❯
      </i>
    </button>
  </div>
</div>
`;

exports[`components/common/Paging > should render last page 1`] = `
.c3 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c2:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 10px;
  border-top: 1px solid #c0c0c0;
}

.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
      title="Předchozí"
      type="button"
    >
      <i
        class="c3"
      >
        ❮
      </i>
    </button>
    <div
      class="c4"
    >
      3
    </div>
    <button
      class="c1 c2"
      disabled=""
      title="Další"
      type="button"
    >
      <i
        class="c3"
        style="color: rgb(192, 192, 192);"
      >
        ❯
      </i>
    </button>
  </div>
</div>
`;

exports[`components/common/Paging > should render one page paging 1`] = `
.c3 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c2:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 10px;
  border-top: 1px solid #c0c0c0;
}

.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
      disabled=""
      title="Předchozí"
      type="button"
    >
      <i
        class="c3"
        style="color: rgb(192, 192, 192);"
      >
        ❮
      </i>
    </button>
    <div
      class="c4"
    >
      1
    </div>
    <button
      class="c1 c2"
      disabled=""
      title="Další"
      type="button"
    >
      <i
        class="c3"
        style="color: rgb(192, 192, 192);"
      >
        ❯
      </i>
    </button>
  </div>
</div>
`;
